### PR TITLE
New midround join screen

### PR DESCRIPTION
**Pull request by AffectedArc07**
## About The Pull Request
This pull request adds in a new join menu, ported from Oracle *F*. It categorises jobs into departments and enlarges the buttons.

No jobs selected:
image

Prioritised Jobs:
image
*Prioritised jobs are show in green bold. HoP is taken because I needed to make them priority*

## Why It's Good For The Game
It looks better than our current menu (below), and is easier to find the job youre looking for
image

## Changelog
:cl:
tweak: The jobs menu has been given a new coat of paint. Yay.
/:cl:

### CODE REVIEW

**Comment by plapatin** [comment]
i was hoping this pr would be a new splash screen

**Comment by nichlas0010** [comment]
https://github.com/tgstation/tgstation/pull/42640 consider looking at this PR and the reviews.

**Comment by AffectedArc07** [comment]
image
image

File: code/modules/mob/dead/new_player/new_player.dm [hunk 1]
```diff
@@ -431,46 +431,81 @@
...
-		var/amt = length(SSjob.prioritized_jobs)
-		var/amt_count
-		for(var/datum/job/a in SSjob.prioritized_jobs)
-			amt_count++
-			if(amt_count != amt) // checks for the last job added.
-				dat += " [a.title], "
-			else
-				dat += " [a.title]. </div>"
+	if(!available_job_count)
+		dat += "<div class='notice red'>There are currently no open positions!</div>"
```
Comment (nichlas0010) [inline code/modules/mob/dead/new_player/new_player.dm]: image

File: code/modules/mob/dead/new_player/new_player.dm [hunk 2]
```diff
@@ -431,46 +431,81 @@
...
-				position_class = "commandPosition"
-			dat += "<a class='[position_class]' href='byond://?src=[REF(src)];SelectedJob=[job.title]'>[job.title] ([job.current_positions])</a><br>"
-	if(!job_count) //if there's nowhere to go, overflow opens up.
+	else
+
+	// if(length(SSjob.prioritized_jobs))
+	// 	dat += "<div class='notice red'>The station has flagged these jobs as high priority:<br>"
+	// 	for(var/datum/job/a in SSjob.prioritized_jobs)
+	// 		dat += " [a.title], "
+	// 	dat += "</div>"
```
Comment (nichlas0010) [inline code/modules/mob/dead/new_player/new_player.dm]: image

File: code/modules/mob/dead/new_player/new_player.dm [hunk 3]
```diff
@@ -431,46 +431,81 @@
...
+			"Miscellaneous" = list(jobs = list(), titles = list(), color = "#ffffff", colBreak = 1),
+			"Synthetic" = list(jobs = list(), titles = GLOB.nonhuman_positions, color = "#ccffcc"),
+			"Service" = list(jobs = list(), titles = GLOB.civilian_positions, color = "#cccccc"),
+			"Medical" = list(jobs = list(), titles = GLOB.medical_positions, color = "#99ffe6", colBreak = 1),
+			"Science" = list(jobs = list(), titles = GLOB.science_positions, color = "#e6b3e6"),
+			"Security" = list(jobs = list(), titles = GLOB.security_positions, color = "#ff9999"),
+		)
 		for(var/datum/job/job in SSjob.occupations)
-			if(job.title != SSjob.overflow_role)
+			if(job && IsJobUnavailable(job.title, TRUE) == JOB_AVAILABLE)
```
Comment (nichlas0010) [inline code/modules/mob/dead/new_player/new_player.dm]: image

File: code/modules/mob/dead/new_player/new_player.dm [hunk 4]
```diff
@@ -431,46 +431,81 @@
...
+								jobs.Insert(1, job)
+							else
+								jobs += job
+						else // Put heads at top of non-command jobs
+							if(job.title in GLOB.command_positions)
+								jobs.Insert(1, job)
+							else
+								jobs += job
+				if(!categorized)
+					categorizedJobs["Miscellaneous"]["jobs"] += job
```
Comment (nichlas0010) [inline code/modules/mob/dead/new_player/new_player.dm]: image

File: code/modules/mob/dead/new_player/new_player.dm [hunk 5]
```diff
@@ -431,46 +431,81 @@
...
 	//src << browse(dat, "window=latechoices;size=300x640;can_close=1")
 
 	// Added the new browser window method
-	var/datum/browser/popup = new(src, "latechoices", "Choose Profession", 440, 500)
+	var/datum/browser/popup = new(src, "latechoices", "Choose Profession", 680, 580)
 	popup.add_stylesheet("playeroptions", 'html/browser/playeroptions.css')
 	popup.set_content(dat)
-	popup.open(FALSE) // FALSE is passed to open so that it doesn't use the onclose() proc
-
+	popup.open(FALSE) // 0 is passed to open so that it doesn't use the onclose() proc
```
Comment (nichlas0010) [inline code/modules/mob/dead/new_player/new_player.dm]: image

**Review by Amelia0010 — CHANGES_REQUESTED** [review]
Since you'd rather just post a picture of the thumbs-ups instead of actually looking at the reviews of the PR

File: code/modules/mob/dead/new_player/new_player.dm (hunk 2 above)
Comment (AffectedArc07) [reply code/modules/mob/dead/new_player/new_player.dm]: As in remove the commented out part?

File: code/modules/mob/dead/new_player/new_player.dm (hunk 4 above)
Comment (AffectedArc07) [reply code/modules/mob/dead/new_player/new_player.dm]: Isnt this where Assistant lies?

File: code/modules/mob/dead/new_player/new_player.dm (hunk 5 above)
Comment (AffectedArc07) [reply code/modules/mob/dead/new_player/new_player.dm]: Will do

File: code/modules/mob/dead/new_player/new_player.dm (hunk 2 above)
Comment (Barhandar) [reply code/modules/mob/dead/new_player/new_player.dm]: yes

File: code/modules/mob/dead/new_player/new_player.dm (hunk 1 above)
Comment (AffectedArc07) [reply code/modules/mob/dead/new_player/new_player.dm]: Will do

File: code/modules/mob/dead/new_player/new_player.dm (hunk 4 above)
Comment (AffectedArc07) [reply code/modules/mob/dead/new_player/new_player.dm]: Wait its under service

File: code/modules/mob/dead/new_player/new_player.dm (hunk 3 above)
Comment (AffectedArc07) [reply code/modules/mob/dead/new_player/new_player.dm]: Ill speak to Jordie on what he fully means by this

**Comment by Tlaltecuhtli** [comment]
pr necromancing

File: code/modules/mob/dead/new_player/new_player.dm (hunk 3 above)
Comment (Jordie0608) [reply code/modules/mob/dead/new_player/new_player.dm]: What do you need clarified?

Comment (AffectedArc07) [reply code/modules/mob/dead/new_player/new_player.dm]: Retrieving `SSjob.occupations.<job>.current_positions`, Im not used to TGs job SS yet so it will take me a while to figure out things

Comment (Jordie0608) [reply code/modules/mob/dead/new_player/new_player.dm]: SSjob has lists of all existing occupations and their datums https://github.com/tgstation/tgstation/blob/master/code/controllers/subsystem/job.dm#L59-L61
So instead of building a list here of the datums just to read their `current_positions` you can directly get it from SSjob with something like
`SSjob.name_occupations["Janitor"].current_positions`

Comment (AffectedArc07) [reply code/modules/mob/dead/new_player/new_player.dm]: So I need to do one of these for each job? Or loop through a set list of all jobs and substitute that in?

I dont know which of these ideas would be better, so apologies for the fuckload of questions

File: code/modules/mob/dead/new_player/new_player.dm [hunk 6]
```diff
@@ -431,47 +432,71 @@
...
-				position_class = "commandPosition"
-			dat += "<a class='[position_class]' href='byond://?src=[REF(src)];SelectedJob=[job.title]'>[job.title] ([job.current_positions])</a><br>"
-	if(!job_count) //if there's nowhere to go, overflow opens up.
+	else
+		dat += "<div class='clearBoth'>Choose from the following open positions:</div><br>"
+		var/list/categorizedJobs = list(
+			"Command" = list(jobs = list(), titles = GLOB.command_positions, color = "#aac1ee"),
+			"Engineering" = list(jobs = list(), titles = GLOB.engineering_positions, color = "#ffd699"),
+			"Supply" = list(jobs = list(), titles = GLOB.supply_positions, color = "#ead4ae", colBreak = 1),
+			"Synthetic" = list(jobs = list(), titles = GLOB.nonhuman_positions, color = "#ccffcc"),
```
Comment (Poojawa) [inline code/modules/mob/dead/new_player/new_player.dm]: `"Ghost Role" = list(jobs = list(), titles = GLOB.mob_spawners, color = "#ffffff"),`

could be placed in the second or third column,

File: code/modules/mob/dead/new_player/new_player.dm [hunk 7]
```diff
@@ -431,47 +432,71 @@
...
+			dat += "<fieldset style='border: 2px solid [color]; display: inline'>"
+			dat += "<legend align='center' style='color: [color]'>[jobcat]</legend>"
+			for(var/datum/job/job in categorizedJobs[jobcat]["jobs"])
+				var/position_class = "otherPosition"
+				if(job.title in GLOB.command_positions)
+					position_class = "commandPosition"
+				if(job in SSjob.prioritized_jobs)
+					dat += "<a class='[position_class]' style='display:block;width:170px' href='byond://?src=[REF(src)];SelectedJob=[job.title]'><font color='lime'><b>[job.title] ([job.current_positions])</b></font></a>"
+				else
+					dat += "<a class='[position_class]' style='display:block;width:170px' href='byond://?src=[REF(src)];SelectedJob=[job.title]'>[job.title] ([job.current_positions])</a>"
```
Comment (Poojawa) [inline code/modules/mob/dead/new_player/new_player.dm]:
```suggestion
					dat += "<a class='[position_class]' style='display:block;width:170px' href='byond://?src=[REF(src)];SelectedJob=[job.title]'>[job.title] ([job.current_positions])</a>"
				categorizedJobs[jobcat]["jobs"] -= job

			for(var/spawner in categorizedJobs[jobcat]["jobs"])
				dat += "<a class='otherPosition' style='display:block;width:170px' href='byond://?src=[REF(src)];JoinAsGhostRole=[spawner]'>[spawner]</a>"
```

Comment (Poojawa) [reply code/modules/mob/dead/new_player/new_player.dm]: if adding Ghost roles to late join menu is wanted.

**Comment by AffectedArc07** [comment]
@optimumtact I can de-stale this if you can eli5 this: https://github.com/tgstation/tgstation/pull/44193#discussion_r289288001

File: code/modules/mob/dead/new_player/new_player.dm (hunk 3 above)
Comment (optimumtact) [reply code/modules/mob/dead/new_player/new_player.dm]: @Jordie0608

Comment (Jordie0608) [reply code/modules/mob/dead/new_player/new_player.dm]: Sorry for the quite late response; I figured it would just be easier to show than tell so I've made you a replacement proc:

```
/mob/dead/new_player/proc/LateChoices()
	var/dat = "<div class='notice'>Round Duration: [DisplayTimeText(world.time - SSticker.round_start_time)]</div>"
	if(SSshuttle.emergency)
		switch(SSshuttle.emergency.mode)
			if(SHUTTLE_ESCAPE)
				dat += "<div class='notice red'>The station has been evacuated.</div><br>"
			if(SHUTTLE_CALL)
				if(!SSshuttle.canRecall())
					dat += "<div class='notice red'>The station is currently undergoing evacuation procedures.</div><br>"
	var/any_jobs_available = FALSE
	for(var/datum/job/job in SSjob.occupations)
		if(job && IsJobUnavailable(job.title, TRUE) == JOB_AVAILABLE)
			any_jobs_available = TRUE
			break
	if(!any_jobs_available)
		dat += "<div class='notice red'>There are currently no open positions!</div>"
	else
		for(var/datum/job/prioritized_job in SSjob.prioritized_jobs)
			if(prioritized_job.current_positions >= prioritized_job.total_positions)
				SSjob.prioritized_jobs -= prioritized_job
		dat += "<table><tr><td valign='top'>"
		for(var/category in GLOB.command_positions + GLOB.engineering_positions + GLOB.supply_positions + GLOB.nonhuman_positions + GLOB.civilian_positions + GLOB.medical_positions + GLOB.science_positions + GLOB.security_positions)
			var/cat_color = SSjob.name_occupations[job].selection_color //use the color of the first job in the category (the department head) as the category color
			dat += "<fieldset style='border: 2px solid [cat_color]; display: inline'>"
			dat += "<legend align='center' style='color: [cat_color]'>[jobcat]</legend>"
			for(var/job in category)
				var/datum/job/job_datum = SSjob.name_occupations[job]
				if(job_datum in SSjob.prioritized_jobs)
					dat += "<[job in GLOB.command_positions ? "class='bold' " : ""]style='display:block;width:170px' href='byond://?src=[REF(src)];SelectedJob=[job_datum.title]'><class='greenannounce'>[job_datum.title] ([job_datum.current_positions])</class></a>"
				else
					dat += "<[job in GLOB.command_positions ? "class='bold' " : ""]style='display:block;width:170px' href='byond://?src=[REF(src)];SelectedJob=[job_datum.title]'>[job_datum.title] ([job_datum.current_positions])</a>"
			dat += "</fieldset><br>"
		dat += "</td></tr></table></center>"
		dat += "</div></div>"
	var/datum/browser/popup = new(src, "latechoices", "Choose Profession", 680, 580)
	popup.set_content(dat)
	popup.open(FALSE) // 0 is passed to open so that it doesn't use the onclose() proc
```

It's quite rough, I didn't even check if it compiled, so you'll probably need to do some editing to make it work.
Note on the use of `selection_color`, the only other place this is used is the character setup occupation preferences so the colors may be a bit different but it'd be good to have that and the latejoin use the same ones, feel free to change both to suit.
I also dropped the use of `var/position_class = "otherPosition"` as a css selector for some existing css classes we have, this means you can delete `html/browser/playeroptions.css` entirely as it'll be unused.

If you have questions ping me in discord, I'm more likely to respond within a week that way =P

**Comment by ExcessiveUseOfCobblestone** [comment]
No ghostrole visibility unless it's config bound, I'm almost certain they don't all have an internal GPS meaning you can get meta info via the joinscreen, nor do I want people to see them comparable to the typical jobs (and demand they be treated as such both administratively and code). They are secondhand roles to placate player's inability to switch servers or in the case of downstreams with only one server, another community entirely upon dying.

File: code/modules/mob/dead/new_player/new_player.dm (hunk 3 above)
Comment (AffectedArc07) [reply code/modules/mob/dead/new_player/new_player.dm]: @Jordie0608 I tried this code, made numerous edits but couldnt get it to smart-pick the category colours, constantly telling me the list was out of bounds. Fir it tried to pull from `job` which isnt defined in this line
`var/cat_color = SSjob.name_occupations[job].selection_color `

I then treid to pull `category[1]` but it kept throwing an out of bounds at me.

**Comment by Poojawa** [comment]
I'm just offering options is all. Ghost roles are solid 'I want to play but not deal with  being on station'

Like how drones were meant to be played

File: code/modules/mob/dead/new_player/new_player.dm (hunk 3 above)
Comment (Jordie0608) [reply code/modules/mob/dead/new_player/new_player.dm]: Yeah I missed quite a few things. When I'm home i'll try actually testing this and message you with an update.

**Comment by AffectedArc07** [comment]
The screen now auto-generates from EXP flags instead of the current snowflake code
image

**Comment by AffectedArc07** [comment]
Holy shit oranges did something good

**Comment by optimumtact** [comment]
orange man gives, orange man takes

**Comment by AffectedArc07** [comment]
fck
